### PR TITLE
Use live illink for trimming tests

### LINKED ISSUE #88624: Update trimming tests to use live-built ILLink.Targets logic
The trimming tests are not using the current ILLink.Targets logic that comes from the dotnet/runtime repo. This is a left-over of the previous versions when this logic came from the dotnet/linker repo. But it is now contained in dotnet/runtime, so we should be using the current repo version so PRs are validated with the updated logic.

              My understanding is that the `trimmingTests` failing here are using the package version picked up in this PR, based on
https://github.com/dotnet/runtime/blob/a41f655468f27325dff62e6a021aa3d05c64e713/eng/testing/linker/project.csproj.template#L55-L62

_Originally posted by @sbomer in https://github.com/dotnet/runtime/issues/88245#issuecomment-1629438201_

### PR DESCRIPTION
**Pull request by sbomer**
Fixes #88624

### CODE REVIEW

**Comment by radical** [comment]
I suggest running `runtime-wasm` also, which runs all the trimming, and AOT tests for wasm.

File: eng/testing/linker/SupportFiles/Directory.Build.targets [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <Project>
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
```
Comment (ViktorHofer) [inline eng/testing/linker/SupportFiles/Directory.Build.targets:2]: That file is expected to imported before the project body is evaluated. Please move this into a Directory.Build.props file so that we build the same as customers do via the PackageReference.

File: src/tools/illink/src/ILLink.RoslynAnalyzer/ILLink.RoslynAnalyzer.csproj [hunk 2]
```diff
@@ -9,6 +9,8 @@
     <AnalysisLevel>Latest</AnalysisLevel>
     <NoWarn Condition="'$(DotNetBuildFromSource)' == 'true'">$(NoWarn);CS8524</NoWarn>
     <AnalyzerLanguage>cs</AnalyzerLanguage>
+    <!-- This analyzer doesn't follow the convention for generator projects so we set the property manually. -->
+    <IsGeneratorProject>true</IsGeneratorProject>
```
Comment (ViktorHofer) [inline src/tools/illink/src/ILLink.RoslynAnalyzer/ILLink.RoslynAnalyzer.csproj:13]: Please undo this change. The `IsGeneratorProject` must be set in a Directory.Build.props file (if not inferred automatically based on the project file name): https://github.com/dotnet/runtime/blob/d4db2d96bb0f55211c30dc69ac587b204bb418f3/Directory.Build.props#L445

File: src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj [hunk 3]
```diff
@@ -20,6 +20,8 @@
          NU5129: Targets don't match the package id. This is intentional. -->
     <NoWarn>$(NoWarn);NU5128;NU5129</NoWarn>
     <IncludeBuildOutput>false</IncludeBuildOutput>
+    <!-- Place build output into a layout that matches the package layout. -->
+    <OutputPath>$(OutputPath)/tools</OutputPath>
```
Comment (ViktorHofer) [inline src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj:24]: The build and package output paths are intentionally not aligned. Why do you need the tools subdirectory for the output path?

File: eng/testing/linker/project.csproj.template [hunk 4]
```diff
@@ -1,5 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
+  <PropertyGroup>
+    <!-- Needed for PublishTrimmed -->
+    <ToolsILLinkDir>{ToolsILLinkDir}</ToolsILLinkDir>
+  </PropertyGroup>
+
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
```
Comment (ViktorHofer) [inline eng/testing/linker/project.csproj.template:8]: Unfortunately this is still not the correct sequence point. The props file is expected to be imported before the project is evaluated, i.e. in the project's Directory.Build.props file. You can achieve that by setting the dependent `ToolsILLinkDir` property before an explicit `Sdk` import and then moving the new `Import` statement into the Directory.Build.props file.

File: src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj (hunk 3 above)
Comment (ViktorHofer) [reply src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj:24]: We should keep the output path and package layout separate, otherwise we will hit other issues over time. Please revert the `OutputPath` change.

File: eng/testing/linker/project.csproj.template [hunk 5]
```diff
@@ -1,4 +1,11 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project>
+
+  <PropertyGroup>
+    <!-- Needed for PublishTrimmed -->
+    <ToolsILLinkDir>{ToolsILLinkDir}</ToolsILLinkDir>
+  </PropertyGroup>
+
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
```
Comment (eerhardt) [inline eng/testing/linker/project.csproj.template:8]: Do we really need to "break out" of using `Sdk="Microsoft.NET.Sdk"` here? Can we just move the `Directory.Build.props` change into this file (so all these "locate local tools" is contained in one place)?

File: eng/pipelines/runtime.yml [hunk 6]
```diff
@@ -288,7 +288,7 @@ extends:
             testGroup: innerloop
             timeoutInMinutes: 120
             nameSuffix: NativeAOT
-            buildArgs: -s clr.aot+host.native+libs -rc $(_BuildConfig) -lc Release -hc Release
+            buildArgs: -s clr.aot+host.native+libs+tools.illink -c $(_BuildConfig) -rc $(_BuildConfig) -lc Release -hc Release
```
Comment (eerhardt) [inline eng/pipelines/runtime.yml:291]: What does -c do when we also specify -rc and -lc?

File: eng/testing/linker/SupportFiles/Directory.Build.props [hunk 7]
```diff
@@ -1,4 +1,13 @@
 <Project>
+
+  <!-- Use live illink bits. It is necessary to both import the package props and override
+       the tasks assembly, because the live package props in the build output do not use
+       the same layout as the NuGet package. -->
+  <Import Project="$(ToolsILLinkDir)build/Microsoft.NET.ILLink.Tasks.props" />
+  <PropertyGroup>
+    <ILLinkTasksAssembly>$(ToolsILLinkDir)net8.0/ILLink.Tasks.dll</ILLinkTasksAssembly>
```
Comment (eerhardt) [inline eng/testing/linker/SupportFiles/Directory.Build.props:8]: Can we use CurrentTargetFramework here? So we don’t have to update net8.0 in net 9 time?

**Review by eerhardt — APPROVED** [review]

File: eng/testing/linker/SupportFiles/Directory.Build.props (hunk 7 above)
Comment (radical) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: What is `CurrentTargetFramework`? I can't find that. I ask because I have an open PR with a similar change - https://github.com/dotnet/runtime/pull/89699 .

Comment (eerhardt) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: https://github.com/dotnet/runtime/blob/30db9b78c8c21aea74a945332bda649e14b6b52d/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/Microsoft.Extensions.Caching.Abstractions.csproj#L4

I meant NetCoreAppCurrent

Comment (vitek-karas) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: I think this should use the `NetCoreAppToolCurrent` because that's what the target project uses: https://github.com/dotnet/runtime/blob/242242947c01615b243e8b4d9ee163030a4ceb47/src/tools/illink/src/ILLink.Tasks/ILLink.Tasks.csproj#L4

File: eng/pipelines/runtime.yml (hunk 6 above)
Comment (ViktorHofer) [reply eng/pipelines/runtime.yml:291]: The -c probably applies to the tools.illink subset.

File: eng/testing/linker/SupportFiles/Directory.Build.props (hunk 7 above)
Comment (ViktorHofer) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: Note that `$(TargetFramework)` shouldn't be used in a props file as that property isn't defined at that time for non multi-targeting projects. You could move this property into the Directory.Build.targets file.

Comment (ViktorHofer) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: Alternatively, you could add the TFM as a property to the imported props file, to keep them in sync.

File: eng/pipelines/runtime.yml (hunk 6 above)
Comment (ViktorHofer) [reply eng/pipelines/runtime.yml:291]: You could simplify this by removing the `-rc` switch.

File: eng/testing/linker/SupportFiles/Directory.Build.props (hunk 7 above)
Comment (vitek-karas) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: I wonder if we should make tools use the runtime configuration. For example when compiling AOT, the "tools" (ilc compiler) use runtime configuration. Additionally I ran into this myself and found it weird that specifying all of `-rc`, `-lc` and `-hc` wasn't enough, that there were still parts which didn't follow either of them.
(Obviously as a separate change)

Comment (ViktorHofer) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: I asked about that when we brought the linker into runtime and the consensus was that the linker should not follow the runtime configuration, as its a separate tool but doesn't need to have yet another configuration property. The current configuration matrix is already quite complicated in dotnet/runtime.

Comment (vitek-karas) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: Do we have other code which falls into the same bucket (as not belonging into any of the 3 configuration properties)?

Comment (ViktorHofer) [reply eng/testing/linker/SupportFiles/Directory.Build.props:8]: Here are the current configuration properties: https://github.com/dotnet/runtime/blob/11d9e87db295032a9927ee19bf3e2918539d8941/eng/Subsets.props#L548-L553

> Do we have other code which falls into the same bucket (as not belonging into any of the 3 configuration properties)?

The shared framework package projects: "packs". But those don't invoke the compiler.

Comment (ViktorHofer) [reply eng/testing/linker/SupportFiles/Directory.Build.props]: > I don't think this works because the imported file ships with the package (so needs to hard code the package TFM) - unless I misunderstood the suggestion.

You already hardcode the TFM in the shipping props file: https://github.com/dotnet/runtime/blob/e43cba31e997e960205844b3f0541011ef732c29/src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.Tasks.props#L21-L22

I was saying that you could introduce another property in that file that just includes the `net8.0\ILLink.Tasks.dll` part that you could then use here to keep the TFM in sync.
